src/events/backup.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Event handlers for creating and restoring backups."""
+import json
+import logging
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.s3 import (
+    CredentialsChangedEvent,
+    CredentialsGoneEvent,
+    S3Requirer,
+)
+from ops import ActionEvent
+from ops.framework import Object
+
+from literals import S3_BACKUPS_PATH, S3_REL_NAME, Status
+from managers.backup import BackupManager
+
+if TYPE_CHECKING:
+    from charm import ZooKeeperCharm
+
+logger = logging.getLogger(__name__)
+
+
+class BackupEvents(Object):
+    """Event handlers for creating and restoring backups."""
+
+    def __init__(self, charm):
+        super().__init__(charm, "backup")
+        self.charm: "ZooKeeperCharm" = charm
+        self.s3_requirer = S3Requirer(self.charm, S3_REL_NAME)
+        self.backup_manager = BackupManager(self.charm.state)
+
+        self.framework.observe(
+            self.s3_requirer.on.credentials_changed, self._on_s3_credentials_changed
+        )
+        self.framework.observe(self.s3_requirer.on.credentials_gone, self._on_s3_credentials_gone)
+
+        self.framework.observe(self.charm.on.create_backup_action, self._on_create_backup_action)
+        self.framework.observe(self.charm.on.list_backups_action, self._on_list_backups_action)
+        self.framework.observe(self.charm.on.restore_action, self._on_restore_action)
+
+    def _on_s3_credentials_changed(self, event: CredentialsChangedEvent):
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.state.peer_relation:
+            self.charm._set_status(Status.NO_PEER_RELATION)
+            event.defer()
+            return
+
+        s3_parameters = self.s3_requirer.get_s3_connection_info()
+        required_parameters = [
+            "bucket",
+            "access-key",
+            "secret-key",
+        ]
+        missing_required_parameters = [
+            param for param in required_parameters if param not in s3_parameters
+        ]
+        if missing_required_parameters:
+            logger.warning(
+                f"Missing required S3 parameters in relation with S3 integrator: {missing_required_parameters}"
+            )
+            self.charm._set_status(Status.MISSING_S3_CONFIG)
+            return
+
+        s3_parameters = self.s3_requirer.get_s3_connection_info()
+
+        # Add some sensible defaults (as expected by the code) for missing optional parameters
+        s3_parameters.setdefault("endpoint", "https://s3.amazonaws.com")
+        s3_parameters.setdefault("region", "")
+        s3_parameters.setdefault("path", S3_BACKUPS_PATH)
+        s3_parameters.setdefault("s3-uri-style", "host")
+        s3_parameters.setdefault("delete-older-than-days", "9999999")
+
+        # Strip whitespaces from all parameters.
```

Comment:
**nitpick:** We can remove this comment.

Reply:
I am usually the one making fun of the comments paraphrasing the code :D 
Removed as part of c463b5232774a7048b2b9963b507ddde54f82f9c.

Though you already gave your approval, one of the changes I made to answer Enrico's review on the second PR is to have a dedicated type module. This is something we talked about recently, so you might want to have a look

Reply:
`core/stubs.py`? If so, I'm totally OK with it.